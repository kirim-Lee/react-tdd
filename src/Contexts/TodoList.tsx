import { createContext, useEffect, useState } from 'react';

interface IContext {
  readonly todoList: string[];
  readonly addTodo: (toDo: string) => void;
  readonly deleteTodo: (index: number) => void;
}

const TodoListContext = createContext<IContext>({
  todoList: [],
  addTodo: () => {},
  deleteTodo: () => {},
});

interface IProps {
  children: JSX.Element | JSX.Element[];
}

const TODO_LIST = 'TodoList';

const TodoListProvider = ({ children }: IProps): JSX.Element => {
  const [todoList, setTodoList] = useState<string[]>([]);

  const addTodo = (toDo: string) => {
    if (!toDo) return;

    const newList = [...todoList, toDo];
    localStorage.setItem(TODO_LIST, JSON.stringify(newList));
    setTodoList(newList);
  };

  const deleteTodo = (index: number) => {
    let list = [...todoList];
    list.splice(index, 1);
    localStorage.setItem(TODO_LIST, JSON.stringify(list));
    setTodoList(list);
  };

  useEffect(() => {
    const list = localStorage.getItem(TODO_LIST);
    if (list) {
      setTodoList(JSON.parse(list));
    }
  }, []);

  return (
    <TodoListContext.Provider value={{ todoList, addTodo, deleteTodo }}>
      {children}
    </TodoListContext.Provider>
  );
};

export { TodoListContext, TodoListProvider };
