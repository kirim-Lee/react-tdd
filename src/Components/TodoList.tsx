import React, { useContext, useState } from 'react';
import styled from 'styled-components';
import { TodoItem } from './TodoItem';
import { TodoListContext } from '../Contexts/TodoList';

const Container = styled.div`
  min-width: 350px;
  height: 400px;
  overflow-y: scroll;
  border: 1px solid #bdbdbd;
`;

export const TodoList = () => {
  const { todoList, deleteTodo } = useContext(TodoListContext);

  return (
    <Container data-testid="toDoList">
      {todoList.map((item, index) => (
        <TodoItem
          key={item}
          id={index}
          label={item}
          onDelete={() => deleteTodo(index)}
        />
      ))}
    </Container>
  );
};
