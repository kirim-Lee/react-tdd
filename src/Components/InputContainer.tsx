import React, { useContext, useState } from 'react';
import styled from 'styled-components';
import { Input } from './Input';
import { Button } from './Button';
import { TodoListContext } from '../Contexts/TodoList';

const Container = styled.div`
  display: flex;
`;

export const InputContainer = () => {
  const { addTodo } = useContext(TodoListContext);
  const [todo, setTodo] = useState<string>('');

  const handleAdd = () => {
    addTodo(todo);
    setTodo('');
  };

  return (
    <Container>
      <Input
        placeholder="할 일을 입력해 주세요"
        value={todo}
        onChange={setTodo}
      />
      <Button label="추가" onClick={handleAdd} />
    </Container>
  );
};
