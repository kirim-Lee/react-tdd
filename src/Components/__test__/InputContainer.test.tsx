import { fireEvent, render, screen } from '@testing-library/react';
import { InputContainer } from '../InputContainer';
import { TodoListProvider } from '../../Contexts/TodoList';

const TODO_LIST = 'TodoList';

describe('<InputContainer />', () => {
  it('렌더링 테스트', () => {
    const { container } = render(<InputContainer />);

    const input = screen.getByPlaceholderText('할 일을 입력해 주세요');
    expect(input).toBeInTheDocument();

    const button = screen.getByText('추가');
    expect(button).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  it('추가버튼을 눌렀을때 값이 초기화 되는지 확인', () => {
    render(<InputContainer />);

    const input = screen.getByPlaceholderText(
      '할 일을 입력해 주세요'
    ) as HTMLInputElement;
    const button = screen.getByText('추가');

    expect(input.value).toBe('');

    fireEvent.change(input, { target: { value: 'study react 1' } });
    expect(input.value).toBe('study react 1');

    fireEvent.click(button);
    expect(input.value).toBe('');
  });

  it('컨텍스트와 연동된 부분을 테스트', () => {
    render(
      <TodoListProvider>
        <InputContainer />
      </TodoListProvider>
    );

    const input = screen.getByPlaceholderText(
      '할 일을 입력해 주세요'
    ) as HTMLInputElement;
    const button = screen.getByText('추가');

    expect(localStorage.getItem(TODO_LIST)).toBeNull();

    fireEvent.change(input, { target: { value: 'study react 1' } });
    fireEvent.click(button);
    expect(localStorage.getItem(TODO_LIST)).toBe(
      JSON.stringify(['study react 1'])
    );
  });

  it('추가 버튼을 눌렀을 때 props으로 전달된 함수 실행하는지 확인', () => {
    const handleClick = jest.fn();
    render(<InputContainer onAdd={handleClick} />);

    const input = screen.getByPlaceholderText(
      '할 일을 입력해 주세요'
    ) as HTMLInputElement;
    const button = screen.getByText('추가');

    expect(handleClick).toHaveBeenCalledTimes(0);

    fireEvent.click(button);

    expect(handleClick).toHaveBeenCalledTimes(0);

    fireEvent.change(input, { target: { value: 'study react 1' } });
    fireEvent.click(button);

    expect(handleClick).toHaveBeenCalledTimes(1);
  });
});
