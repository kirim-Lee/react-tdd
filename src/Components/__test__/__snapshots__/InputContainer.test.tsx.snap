// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InputContainer /> 렌더링 테스트 1`] = `
<div>
  <div
    class="sc-hLBbgP gJsMjZ"
  >
    <input
      class="sc-bcXHqe dKujtN"
      placeholder="할 일을 입력해 주세요"
      value=""
    />
    <button
      class="sc-gswNZR cMvRBc"
    >
      <div
        class="sc-dkrFOg dHCelr"
      >
        추가
      </div>
    </button>
  </div>
</div>
`;
